diff --git a/dist/index.js b/dist/index.js
index ce328f1bc67710ca5c3bf303d552ba7be0b87a2e..da33c26eb65b3fea197ca38218fa3bb7d4e01ddb 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -285,7 +285,7 @@ var DialogContentImpl = React.forwardRef(
         }
       ),
       /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
-        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TitleWarning, { titleId: context.titleId }),
+        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TitleWarning, { titleId: context.titleId, contentRef }),
         /* @__PURE__ */ (0, import_jsx_runtime.jsx)(DescriptionWarning, { contentRef, descriptionId: context.descriptionId })
       ] })
     ] });
@@ -335,7 +335,7 @@ var [WarningProvider, useWarningContext] = (0, import_react_context.createContex
   titleName: TITLE_NAME,
   docsSlug: "dialog"
 });
-var TitleWarning = ({ titleId }) => {
+var TitleWarning = ({ titleId, contentRef }) => {
   const titleWarningContext = useWarningContext(TITLE_WARNING_NAME);
   const MESSAGE = `\`${titleWarningContext.contentName}\` requires a \`${titleWarningContext.titleName}\` for the component to be accessible for screen reader users.
 
@@ -344,7 +344,8 @@ If you want to hide the \`${titleWarningContext.titleName}\`, you can wrap it wi
 For more information, see https://radix-ui.com/primitives/docs/components/${titleWarningContext.docsSlug}`;
   React.useEffect(() => {
     if (titleId) {
-      const hasTitle = document.getElementById(titleId);
+      const root = contentRef?.current?.getRootNode() ?? document;
+      const hasTitle = root.getElementById ? root.getElementById(titleId) : document.getElementById(titleId);
       if (!hasTitle) console.error(MESSAGE);
     }
   }, [MESSAGE, titleId]);
@@ -357,7 +358,8 @@ var DescriptionWarning = ({ contentRef, descriptionId }) => {
   React.useEffect(() => {
     const describedById = contentRef.current?.getAttribute("aria-describedby");
     if (descriptionId && describedById) {
-      const hasDescription = document.getElementById(descriptionId);
+      const root = contentRef.current?.getRootNode() ?? document;
+      const hasDescription = root.getElementById ? root.getElementById(descriptionId) : document.getElementById(descriptionId);
       if (!hasDescription) console.warn(MESSAGE);
     }
   }, [MESSAGE, contentRef, descriptionId]);
diff --git a/dist/index.mjs b/dist/index.mjs
index 13581033b5a5635796d4cf492f0828e853d011c4..12b4794fe38af47cd466f3b1137d0d35e3a2085f 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -233,7 +233,7 @@ var DialogContentImpl = React.forwardRef(
         }
       ),
       /* @__PURE__ */ jsxs(Fragment, { children: [
-        /* @__PURE__ */ jsx(TitleWarning, { titleId: context.titleId }),
+        /* @__PURE__ */ jsx(TitleWarning, { titleId: context.titleId, contentRef }),
         /* @__PURE__ */ jsx(DescriptionWarning, { contentRef, descriptionId: context.descriptionId })
       ] })
     ] });
@@ -283,7 +283,7 @@ var [WarningProvider, useWarningContext] = createContext(TITLE_WARNING_NAME, {
   titleName: TITLE_NAME,
   docsSlug: "dialog"
 });
-var TitleWarning = ({ titleId }) => {
+var TitleWarning = ({ titleId, contentRef }) => {
   const titleWarningContext = useWarningContext(TITLE_WARNING_NAME);
   const MESSAGE = `\`${titleWarningContext.contentName}\` requires a \`${titleWarningContext.titleName}\` for the component to be accessible for screen reader users.
 
@@ -292,7 +292,8 @@ If you want to hide the \`${titleWarningContext.titleName}\`, you can wrap it wi
 For more information, see https://radix-ui.com/primitives/docs/components/${titleWarningContext.docsSlug}`;
   React.useEffect(() => {
     if (titleId) {
-      const hasTitle = document.getElementById(titleId);
+      const root = contentRef?.current?.getRootNode() ?? document;
+      const hasTitle = root.getElementById ? root.getElementById(titleId) : document.getElementById(titleId);
       if (!hasTitle) console.error(MESSAGE);
     }
   }, [MESSAGE, titleId]);
@@ -305,7 +306,8 @@ var DescriptionWarning = ({ contentRef, descriptionId }) => {
   React.useEffect(() => {
     const describedById = contentRef.current?.getAttribute("aria-describedby");
     if (descriptionId && describedById) {
-      const hasDescription = document.getElementById(descriptionId);
+      const root = contentRef.current?.getRootNode() ?? document;
+      const hasDescription = root.getElementById ? root.getElementById(descriptionId) : document.getElementById(descriptionId);
       if (!hasDescription) console.warn(MESSAGE);
     }
   }, [MESSAGE, contentRef, descriptionId]);
